### ISSUE TITLE
ServerExceptionMapper not called

### ISSUE THREAD

**Issue by mschirmacher** [quarkusio/quarkus#44945]
### Describe the bug

ServerExceptionMappers are supposed to be always consulted if they perfectly match the exception.
But if proactive authentication is turned on, they are not.

### Expected behavior

Customt ExceptionMapper should be called if authentication fails

### Actual behavior

they are not.

### How to Reproduce?

run this [test](https://github.com/mschirmacher/QuarkusAuthorizationPolicySuspendBug/blob/noExceptionMapperCalled/src/test/kotlin/org/acme/GreetingResourceTest.kt)
If the custom exception mapper is called, the test should be green.
Note the config on line 29 of the test - if set to false the test is green!

Architectural overview:
The test sets up a wiremock response, that returns invalid json.
There is a [custom identity provider](https://github.com/mschirmacher/QuarkusAuthorizationPolicySuspendBug/blob/noExceptionMapperCalled/src/main/kotlin/org/acme/security/auth/CustomIdentityProvider.kt), that will call an external service (the one mocked by wiremock) for authentication.
As the response cannot be deserialized, an exception is thrown which SHOULD be handled by the matching [ServerExeptionMapper](https://github.com/mschirmacher/QuarkusAuthorizationPolicySuspendBug/blob/noExceptionMapperCalled/src/main/kotlin/org/acme/exception/ServerExceptionMapper.kt) - but its not.

### Additional information

(my unprofessional debugging attemts reveald, that [this]{io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder} class is the bad guy as it doesnt seem to be prepared to call custom exception mappers)

**Comment by geoand** [comment]
When I run the test I get:

```posh
java.lang.AssertionError: 1 expectation failed.
Expected status code <234> but was <500>.

	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:73)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:108)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:57)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure.validate(ResponseSpecificationImpl.groovy:512)
	at io.restassured.internal.ResponseSpecificationImpl$HamcrestAssertionClosure$validate$1.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at io.restassured.internal.ResponseSpecificationImpl.validateResponseIfRequired(ResponseSpecificationImpl.groovy:696)
	at io.restassured.internal.ResponseSpecificationImpl.this$2$validateResponseIfRequired(ResponseSpecificationImpl.groovy)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:198)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:62)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:135)
	at io.restassured.specification.ResponseSpecification$statusCode$0.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at io.restassured.internal.ResponseSpecificationImpl.statusCode(ResponseSpecificationImpl.groovy:143)
	at io.restassured.internal.ValidatableResponseOptionsImpl.statusCode(ValidatableResponseOptionsImpl.java:89)
	at org.acme.GreetingResourceTest.testHelloEndpoint(GreetingResourceTest.kt:53)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:967)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:817)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: org.jboss.resteasy.reactive.ClientWebApplicationException: Received: 'HTTP 200 OK' when invoking REST Client method: 'org.acme.security.auth.ExternalServiceClient#getTokenInfo'
		at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.unwrapException(RestClientRequestContext.java:205)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.handleException(AbstractResteasyReactiveContext.java:331)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:175)
		at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext$1.lambda$execute$0(RestClientRequestContext.java:324)
		at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
		at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
		at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:50)
		at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
		at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
		at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:1583)
	Caused by: com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'invalid': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 9]
		at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:2584)
		at com.fasterxml.jackson.core.JsonParser._constructReadException(JsonParser.java:2610)
		at com.fasterxml.jackson.core.JsonParser._constructReadException(JsonParser.java:2618)
		at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:825)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._reportInvalidToken(UTF8StreamJsonParser.java:3662)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._handleUnexpectedValue(UTF8StreamJsonParser.java:2749)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._nextTokenNotInObject(UTF8StreamJsonParser.java:867)
		at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:753)
		at com.fasterxml.jackson.databind.ObjectReader._initForReading(ObjectReader.java:357)
		at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2115)
		at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1501)
		at io.quarkus.rest.client.reactive.jackson.runtime.serialisers.ClientJacksonMessageBodyReader.readFrom(ClientJacksonMessageBodyReader.java:51)
		at org.jboss.resteasy.reactive.client.impl.ClientReaderInterceptorContextImpl.proceed(ClientReaderInterceptorContextImpl.java:86)
		at org.jboss.resteasy.reactive.client.impl.ClientSerialisers.invokeClientReader(ClientSerialisers.java:160)
		at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.readEntity(RestClientRequestContext.java:219)
		at org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler.mapToResponse(ClientResponseCompleteRestHandler.java:113)
		at org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler.handle(ClientResponseCompleteRestHandler.java:35)
		at org.jboss.resteasy.reactive.client.handlers.ClientResponseCompleteRestHandler.handle(ClientResponseCompleteRestHandler.java:31)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.invokeHandler(AbstractResteasyReactiveContext.java:231)
		at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
		... 12 more

```


which tells me that:

```kotlin
@RegisterRestClient(configKey = "foo")
interface ExternalServiceClient {

    @GET
    @Path("/{token}")
    fun getTokenInfo(
        @PathParam("token") token: String
    ): Uni<TokenInfo>
}

data class TokenInfo(val isValid: Boolean)

```

is not properly implemented.

As this is an exception originating from the REST Client, `@ServerExceptionMapper` is not part of picture

**Comment by mschirmacher** [comment]
@geoand yes, that is the expected outcome.
And yes, the error comes from the rest client. BUT if you disable proactive authentication, you will find the test is green.

**Comment by mschirmacher** [comment]
That's not true. Also without proactive authentication the `CustomIdentityProvider` is used.
Otherwise the test would still fail

**Comment by geoand** [comment]
Seems to indeed be the case - I was tricked being tricked by the debugger.

At this point I need @sberyozkin to look into it to see why the security stuff affects this case

**Comment by sberyozkin** [comment]
@mschirmacher 

>And yes, the error comes from the rest client. BUT if you disable proactive authentication, you will find the test is green.

By default, when the proactive authentication is enabled, the authentication is enforced before the JAX-RS chain runs, but Michal made sure that even in this case JAX-RS exception mappers are called for `AuthenticationException` exception instances.

When the proactive authentication is disabled, the JAX-RS chain is active while the authentication is enforced.

I'd not qualify this issue as a bug, it has never been an expectation that a security chain in the proactive authentication mode can support mapping of random JAX-RS exceptions to exception mappers, it looks out of scope for the security layer which is triggered before the JAX-RS chain even starts.

Perhaps something can be tweaked in this case to try to connect random exceptions to JAX-RS exception mappers, @michalvavrik, do you think it is reasonable ? Or should we ask users to disable the proactive authentication ?

**Comment by michalvavrik** [comment]
I agree with @sberyozkin summary here https://github.com/quarkusio/quarkus/issues/44945#issuecomment-2520915147.

> Perhaps something can be tweaked in this case to try to connect random exceptions to JAX-RS exception mappers, @michalvavrik, do you think it is reasonable ? Or should we ask users to disable the proactive authentication ?

Implementation wise it is not hard to loosen this, but despite what it seems to be, this is not just security matter. Let me try to explain it:

- The mechanism that propagates exceptions from Vert.x failure handlers to Quarkus REST has been added explicitly by us (Sergey and me) to allow handling of authentication exceptions with exception mappers because a lot of users wanted it. Part of these users basically used it for logging etc. which they can now do with security events.
- We intentionally added condition that it is only propagated for auth exceptions otherwise any route failure unrelated to Quarkus REST (before Quarkus REST starts) would result in starting Quarkus REST just in case users want to catch it. I think it would result in  a lot more processing that users generally don't ask for.

Conclusion: I didn't run reproducer yet, but if I understand situation correctly, REST client throws exception inside identity provider. @sberyozkin IMO it should be authentication failure anyway, WDYT? I think we should wrap this. It should also ensure correct HTTP status will be returned, right? If you agree, I'll do it tomorrow, but please bear in mind I didn't check reproducer yet. Thanks

**Comment by michalvavrik** [comment]
If it is not clear, I suggest to wrap the exception returned from the identity provider, which means you will be able to use the exception mapper. I suppose you can use `@Unwrap` (something like that exist, I remember Georgios added it, but don't remember exact annotation name).

**Comment by geoand** [comment]
> REST client throws exception inside identity provider

Yeah, this is what is happening

**Comment by sberyozkin** [comment]
Hi @michalvavrik, @geoand  Indeed, wrapping them up for the users to check causes in the mappers makes sense 👍 .

Michal, just for the record, you did all the hard work, I might've added a few comments only :-)

Hmm... thinking about it a bit more, returning `401` by wrapping an unrecognized exception into `AuthenticationFailedException`, which would otherwise cause `500`, might be a bit confusing if the user provided correct credentials but the identity provider can't process them due to some of its internal failures... Actually, something similar might happen in some OIDC cases...

How about introducing an exception class like `GeneralAuthenticationException`  or something like that, wrap unrecognized exceptions into this one, if the exception mappers catch it - they can unwrap it and decide on the status, otherwise it will eventually produce 500... Or is it too messy ?

**Comment by sberyozkin** [comment]
>How about introducing an exception class like GeneralAuthenticationException or something like that, wrap unrecognized exceptions into this one, if the exception mappers catch it - they can unwrap it and decide on the status, otherwise it will eventually produce 500... Or is it too messy ?

Ignore it please, users can do exactly the same with `AuthenticationFailedException`

**Comment by michalvavrik** [comment]
> Hmm... thinking about it a bit more, returning 401 by wrapping an unrecognized exception into AuthenticationFailedException, which would otherwise cause 500, might be a bit confusing if the user provided correct credentials but the identity provider can't process them due to some of its internal failures...

500 response status gives away something is wrong. But it can also be something wrong based on incoming HTTP request (some header, input, cookie, whatever) and I don't think we should provide this information in a prod mode by default. Users can map this to their response status if they want. In DEV/TEST mode, we need to return clear exception or logging. I think it is not relevant for actual user whether it is 500 or 401, it is only relevant for maintainer/developer and they need to use logs.

> Actually, something similar might happen in some OIDC cases...

I know about it and thought about it many times for reasons/opinions in mentioned above ^^.

**Comment by geoand** [comment]
> I think it is not relevant for actual user whether it is 500 or 401 (because how it helps him??), it is only relevant for maintainer/developer and they need to use logs.

I would like the opinion of @FroMage here as well

**Comment by FroMage** [comment]
> I would like the opinion of @FroMage here as well

My opinion is that we should nerd-snipe @gsmet to make a github bot that reacts on "let's ask Mr Foo's opinion" and asks ChatGPT to summarise the discusion and the points on which you want Mr Foo's opinion 🤣

Failing that, let me try to get this straight:

- We have a server endpoint (I assume Quarkus REST is involved otherwise we would not be talking about `@ServerExceptionMapper`) that is protected by auth.
- Auth layer calls the Quarkus REST Client to figure out whether auth is granted (this is not called explicitely by the Quarkus REST endpoint).
- That REST Client call fails due to a JSON error `JsonParseException` (Confusing too… `ClientWebApplicationException: Received: 'HTTP 200 OK'`)
- In the case of non-proactive auth, the exception is correctly mapped via `@ServerExceptionMapper` because it's set up in the chain
- In the case of proactive auth, the exception is not mapped because our security layer will only map `AuthenticationException` and this is not its type
- There appears to be an additional question of whether to return a 500 or 401 in this case

So, let's see:

- It's nice that proactive auth calls server exception mappers, but it's pretty confusing that it only does so for a single type of exception. That's hard to explain, _unless_ that is the only kind of exception that proactive auth can throw
- Authentication apparently can throw more sorts of errors than "authentication was not granted after we successfully checked". It can also have "internal errors" that were not expected (the case here: a JSON error) but also cause authentication to fail, as far as the request is concerned.
- In both "it worked, not granted" and "something went wrong, not granted", the request caller will not be able to complete the request, and we could argue that it's a 401 in both cases, and that the request caller is unconcerned about why, however:
- There is a difference in status codes between 401 and 500, which is crucial, which is that in the first case it's the user's fault, and they will wonder why it happens and check that they're logged in properly, and perhaps check their user, roles, permissions, and assume they can fix the problem. Whlie in the case of 500, it's the server's at fault, and the user does not have to waste time trying to fix their request.
- So in my opinion, this is clearly a 500: the server has a bug that must be fixed.
- In both cases: proactive, or non-proactive.
- If we agree on that (and perhaps I'm wrong and missing something), then we could argue that it's fine for proactive auth to only use exception mappers for `AuthenticationException` (check went well, access not granted) and leave the rest as 500, but…
- We can also argue that we should allow other kinds of exceptions to be mapped in proactive mode, since we allow 500 to be mapped in non-proactive mode.

I do get the argument that we're not sure it's a good idea to fire up Quarkus REST in the case proactive auth just for exception mapping, especially because the request, had it gone through, might not even end up being handled by Quarkus REST.

But it's pretty hard to defend only calling mappers for some kinds of exceptions.

**Comment by geoand** [comment]
> My opinion is that we should nerd-snipe @gsmet to make a github bot that reacts on "let's ask Mr Foo's opinion" and asks ChatGPT to summarise the discusion and the points on which you want Mr Foo's opinion 🤣

That's actually not hard to do 😉

**Comment by michalvavrik** [comment]
I think @FroMage summary is correct and answers are good, I just have bit different preference, that's all.

> There is a difference in status codes between 401 and 500, which is crucial, which is that in the first case it's the user's fault, and they will wonder why it happens and check that they're logged in properly, and perhaps check their user, roles, permissions, and assume they can fix the problem. Whlie in the case of 500, it's the server's at fault, and the user does not have to waste time trying to fix their request.

Possibly, just because it results in 500 once, it doesn't mean you shouldn't try again unless you know that result is idempotent. Depending on how it's written, you can stop on the client side, but you always need fault tolerance on the server side in case someone keeps trying for a good reason or not.

> So in my opinion, this is clearly a 500: the server has a bug that must be fixed.

but that can take  a time depending on SLAs etc. I am not sure if users should only start debugging if 5xx is returned and not watch logs etc. for warns/errors. It doesn't seem like people are concerned with giving away this information, so I'll respect that's the way we go (keep 5xx).

> If we agree on that (and perhaps I'm wrong and missing something), then we could argue that it's fine for proactive auth to only use exception mappers for AuthenticationException (check went well, access not granted) and leave the rest as 500, but…
We can also argue that we should allow other kinds of exceptions to be mapped in proactive mode, since we allow 500 to be mapped in non-proactive mode.

I am fine with propagating exceptions thrown during authentication/authorization before Quarkus REST/RESTEasy had started, but no others. I don't know about other failure handlers extensions/users can define and don't want to catch it all. It is easy to do as I said, we can introduce some wrapper exception that we unwrap in RESTEasy/REST before they start processing. I can do that if others agree with @FroMage proposal.

**Comment by geoand** [comment]
> I am fine with propagating exceptions thrown during authentication/authorization before Quarkus REST/RESTEasy had started, but no others. I don't know about other failure handlers extensions/users can define and don't want to catch it all. It is easy to do as I said, we can introduce some wrapper exception that we unwrap in RESTEasy/REST before they start processing.

+1

**Comment by mschirmacher** [comment]
Hello everybody. 
Thank you for this discussion :-)
I want to throw in another point:
You discussed, if it should be a 401 or a 500 (i agree, in the issue i reported, 500 would be adequate - clear problem on the server side, while the user made made an a valid request).
BUT its not just about the status code, but also about the response body.

There is an api spec describing the service im implementing and i must be able to control the response body, in order to adhere to that spec.
It would be inconvenient on the one hand, to have a special exception handler for authentication, and  IMHO far from intuitive  OTOH.
Yes, the docs say you CAN handle `AuthenticationException`, but from reading it, it is not clear, that this is the only way to handle exception.
Also it is not clear, you have to catch all exception with try/catch, just to wrap them in AuthenticationException just in order to handle it with the defined handler...

**Comment by michalvavrik** [comment]
Thanks for your view on the matter @mschirmacher, it will be fixed in 3.18, there is already PR. Fix probably shouldn't be backported, I'd like to be careful as in the past, it took us a couple of iterations to get this propagation of exceptions right.

**Comment by michalvavrik** [comment]
I am not sure if you know, but I have already linked this documentation https://quarkus.io/guides/security-proactive-authentication#customize-auth-exception-responses and it shows how to define ` failureHandler` that should serve you as a workaround till the 3.18 is release. It also works for other than authentication exceptions.
